src/transform/src/dataflow.rs

```diff
@@ -434,76 +449,382 @@ pub fn optimize_dataflow_monotonic(dataflow: &mut DataflowDesc) -> Result<(), Tr
+    // (currently only theoretical) possibility that we are using an index that we are also building
+    // ourselves.
+    for (_id, sink_desc) in dataflow.sink_exports.iter() {
+        index_reqs_by_id
+            .entry(sink_desc.from)
+            .or_insert_with(Vec::new);
     }
-    for (_, (index_desc, _)) in dataflow.index_exports.iter() {
-        indexes_by_view
+    for (_id, (index_desc, _)) in dataflow.index_exports.iter() {
+        index_reqs_by_id
             .entry(index_desc.on_id)
-            .or_insert_with(BTreeSet::new);
+            .or_insert_with(Vec::new);
     }
+
+    // Go through the MIR plans of `objects_to_build` and collect which arrangements are requested
+    // for which we also have an available index.
     for build_desc in dataflow.objects_to_build.iter_mut() {
-        build_desc
-            .plan
-            .as_inner_mut()
-            .visit_post(&mut |e| match e {
-                MirRelationExpr::Get {
-                    id: Id::Global(id), ..
-                } => {
-                    indexes_by_view.entry(*id).or_insert_with(BTreeSet::new);
-                }
-                MirRelationExpr::ArrangeBy { input, keys } => {
-                    if let MirRelationExpr::Get {
-                        id: Id::Global(id), ..
-                    } = &**input
-                    {
-                        for key in keys {
-                            if indexes.indexes_on(*id).find(|k| k == &key).is_some() {
-                                indexes_by_view.get_mut(id).unwrap().insert(key.clone());
-                            }
-                        }
-                    }
-                }
-                _ => {}
-            })?;
+        CollectIndexRequests::new(indexes, &mut index_reqs_by_id)
+            .collect_index_reqs(build_desc.plan.as_inner_mut())?;
     }
-    for (id, keys) in indexes_by_view.iter_mut() {
-        if keys.is_empty() {
-            // If a dataflow uses a view that has indexes, but it has not
-            // chosen to use any specific index of that view, pick an arbitrary
-            // index to read from.
+
+    // Add full index scans, i.e., where an index exists, but either no specific key was requested
+    // or the requested key didn't match an existing index in the above code.
+    for (id, requests_with_matching_indexes) in index_reqs_by_id.iter_mut() {
+        if requests_with_matching_indexes.is_empty() {
+            // Pick an arbitrary index to be fully scanned.
+            // TODO: There are various edge cases where a better choice would be possible:
+            // - Some indexes might be less skewed than others.
+            // - Some indexes might have an error, while others don't.
+            //   https://github.com/MaterializeInc/materialize/issues/15557
+            // - Some indexes might have less extra data in their keys, which won't be used in a
+            //   full scan.
```

Comment:
Not super relevant to the PR, but: what would we need to know to make an effective cost-based decision here?